sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala

```diff
@@ -112,27 +98,46 @@ object GpuSemaphore {
   }
 }
 
-private final class GpuSemaphore(tasksPerGpu: Int) extends Logging with Arm {
-  private val semaphore = new Semaphore(tasksPerGpu)
+private final class GpuSemaphore() extends Logging with Arm {
+  private val maxPermits = 1000
+  private val semaphore = new Semaphore(maxPermits)
 
   // Map to track which tasks have acquired the semaphore.
-  case class TaskInfo(count: MutableInt, thread: Thread)
+  case class TaskInfo(count: MutableInt, thread: Thread, numPermits: Int)
   private val activeTasks = new ConcurrentHashMap[Long, TaskInfo]
 
+  def computeNumPermits(conf: SQLConf): Int = {
+    val concurrentStr = conf.getConfString(RapidsConf.CONCURRENT_GPU_TASKS.key, null)
+    val concurrentInt = Option(concurrentStr)
+      .map(ConfHelper.toInteger(_, RapidsConf.CONCURRENT_GPU_TASKS.key))
+      .getOrElse(RapidsConf.CONCURRENT_GPU_TASKS.defaultValue)
+    val permits = maxPermits / concurrentInt
```

Comment:
should we guard against `concurrentGpuTasks` being 0? I know we could do that in the past.. but here we can get a divide-by-zero exception. I mean constraining the config with `checkValue`